diff --git a/meson.build b/meson.build
index 0496aaba50d..47bfc297485 100644
--- a/meson.build
+++ b/meson.build
@@ -1856,7 +1856,7 @@ endif
 
 dep_clang = null_dep
 if with_clc or with_gallium_clover
-  llvm_libdir = dep_llvm.get_variable(cmake : 'LLVM_LIBRARY_DIR', configtool: 'libdir')
+  llvm_libdir = get_option('clang-libdir')
 
   dep_clang = cpp.find_library('clang-cpp', dirs : llvm_libdir, required : false)
 
diff --git a/meson_options.txt b/meson_options.txt
index 79ee65e6094..9d80e097463 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -1,6 +1,13 @@
 # Copyright © 2017-2019 Intel Corporation
 # SPDX-License-Identifier: MIT
 
+option(
+  'clang-libdir',
+  type : 'string',
+  value : '',
+  description : 'Locations to search for clang libraries.'
+)
+
 option(
   'split-debug',
   type : 'feature',
diff --git a/src/gallium/targets/opencl/meson.build b/src/gallium/targets/opencl/meson.build
index ab2c83556a8..a959441cb36 100644
--- a/src/gallium/targets/opencl/meson.build
+++ b/src/gallium/targets/opencl/meson.build
@@ -56,8 +56,7 @@ if with_opencl_icd
     configuration : _config,
     input : 'mesa.icd.in',
     output : 'mesa.icd',
-    install : true,
-    install_tag : 'runtime',
+    install : false,
     install_dir : join_paths(get_option('sysconfdir'), 'OpenCL', 'vendors'),
   )
 
diff --git a/src/gallium/targets/rusticl/meson.build b/src/gallium/targets/rusticl/meson.build
index 2b214adbbf5..53e466174d5 100644
--- a/src/gallium/targets/rusticl/meson.build
+++ b/src/gallium/targets/rusticl/meson.build
@@ -64,8 +64,7 @@ configure_file(
   configuration : _config,
   input : 'rusticl.icd.in',
   output : 'rusticl.icd',
-  install : true,
-  install_tag : 'runtime',
+  install : false,
   install_dir : join_paths(get_option('sysconfdir'), 'OpenCL', 'vendors'),
 )
 
