diff --git a/rust/kernel/lib.rs b/rust/kernel/lib.rs
index f56e077b8f53..5e0625b41dad 100644
--- a/rust/kernel/lib.rs
+++ b/rust/kernel/lib.rs
@@ -23,6 +23,7 @@
 #![feature(type_alias_impl_trait)]
 #![feature(unsize)]
 #![warn(clippy::undocumented_unsafe_blocks)]
+#![feature(box_uninit_write)]
 
 // Ensure conditional compilation based on the kernel configuration works;
 // otherwise we may silently break things like initcall handling.
