diff --git a/rust/alloc/alloc.rs b/rust/alloc/alloc.rs
index ca224a541..dac9e96d3 100644
--- a/rust/alloc/alloc.rs
+++ b/rust/alloc/alloc.rs
@@ -30,13 +30,13 @@ extern "Rust" {
     // The rustc fork of LLVM also special-cases these function names to be able to optimize them
     // like `malloc`, `realloc`, and `free`, respectively.
     #[rustc_allocator]
-    #[rustc_allocator_nounwind]
+    #[rustc_nounwind]
     fn __rust_alloc(size: usize, align: usize) -> *mut u8;
-    #[rustc_allocator_nounwind]
+    #[rustc_nounwind]
     fn __rust_dealloc(ptr: *mut u8, size: usize, align: usize);
-    #[rustc_allocator_nounwind]
+    #[rustc_nounwind]
     fn __rust_realloc(ptr: *mut u8, old_size: usize, align: usize, new_size: usize) -> *mut u8;
-    #[rustc_allocator_nounwind]
+    #[rustc_nounwind]
     fn __rust_alloc_zeroed(size: usize, align: usize) -> *mut u8;
 }
 
diff --git a/rust/alloc/boxed.rs b/rust/alloc/boxed.rs
index 8fd296421..fa72125e8 100644
--- a/rust/alloc/boxed.rs
+++ b/rust/alloc/boxed.rs
@@ -1972,7 +1972,7 @@ impl<T: ?Sized, A: Allocator> AsMut<T> for Box<T, A> {
  */
 #[stable(feature = "pin", since = "1.33.0")]
 #[rustc_const_unstable(feature = "const_box", issue = "92521")]
-impl<T: ?Sized, A: Allocator> const Unpin for Box<T, A> where A: 'static {}
+impl<T: ?Sized, A: Allocator> Unpin for Box<T, A> where A: 'static {}
 
 #[unstable(feature = "generator_trait", issue = "43122")]
 impl<G: ?Sized + Generator<R> + Unpin, R, A: Allocator> Generator<R> for Box<G, A>
diff --git a/rust/kernel/driver.rs b/rust/kernel/driver.rs
index 7154c0221..af42a7732 100644
--- a/rust/kernel/driver.rs
+++ b/rust/kernel/driver.rs
@@ -119,6 +119,7 @@ impl<T: DriverOps> Drop for Registration<T> {
 ///   - [`RawDeviceId::ZERO`] is actually a zeroed-out version of the raw device id.
 ///   - [`RawDeviceId::to_rawid`] stores `offset` in the context/data field of the raw device id so
 ///     that buses can recover the pointer to the data.
+#[const_trait]
 pub unsafe trait RawDeviceId {
     /// The raw type that holds the device id.
     ///
